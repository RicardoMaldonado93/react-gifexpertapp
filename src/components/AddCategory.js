import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({setCategory}) => {
    
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = ({target}) =>setInputValue(target.value);

    const handleSubmit = (e)=>{ 
        e.preventDefault(); 

        if(inputValue.trim().length > 2){
            setCategory( items =>[  inputValue, ...items ]);
            setInputValue("");
            // document.querySelector("input").placeholder = "ingrese categoria..."
        }

    };

    return (
        <form onSubmit={handleSubmit}>
            {/*  A modo de test unitario para hacer pruebas de value  */}
            <p style={{'opacity':'0.01', 'position':'absolute'}}>{inputValue}</p>
            <input 
                type="text"
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )

}

AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired
}

