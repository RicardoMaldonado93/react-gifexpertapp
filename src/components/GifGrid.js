import React from 'react'
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {
    const { data:images, loading } = useFetchGif( category );

    return (
        <>
            <h3 className="animate__animated animate__fadeIn"> ~ { category } ~ </h3>
            { loading && <p className="animate__animated animate__flash">cargando...</p> }
            <div className="card-grid animate__animated animate__fadeIn">
                {         
                    images.map( img =>
                    <GifGridItem 
                        {...img}
                        key={img.id}
                    />  
                    )
                }
            </div>
        </>
    )
}
