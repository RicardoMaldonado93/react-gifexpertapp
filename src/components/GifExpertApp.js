import React, { useState } from 'react'
import { AddCategory } from './AddCategory';
import { GifGrid } from './GifGrid';

const GifExpertApp = ({ defaultCategories = [] })=>{
    
    const [categories, setCategories] = useState(defaultCategories);

    return (
        <>
            <h2>GifExpertApp</h2>
            <AddCategory setCategory={ setCategories }/>
            <hr />

            <ol>
                {
                    categories.map( category =>
                        <GifGrid 
                            category={category}
                            key={category}
                        />
                    )
                }

            </ol>
            
        </>
    );
}

export default GifExpertApp;