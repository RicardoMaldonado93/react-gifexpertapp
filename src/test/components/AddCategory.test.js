import '@testing-library/jest-dom'
import React from 'react';
import { shallow } from "enzyme";
import { AddCategory } from '../../components/AddCategory';

describe("Unit test for 'AddCategory'", () => {
    const setCategory = jest.fn();
    let wrapper = shallow(<AddCategory setCategory={setCategory}/>);
    
    beforeEach(()=>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategory={setCategory}/>);
    });

    test("should render correctly component", () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should change text input', () => {
        const input = wrapper.find("input");
        const value = "Hola mundo";
        
        input.simulate("change",{ target:{value} });
        const p = wrapper.find('p');
        
        expect( p.text().trim() ).toBe(value);
    });

    test("'No' post the information with submit", () => {
        wrapper.find("form").simulate("submit",{ preventDefault(){} });
        expect( setCategory ).not.toHaveBeenCalled();
    });

    test("should call setCategories and clean input",()=>{
        const value = "one punch";

        //* Simulate change 
        wrapper.find("input").simulate("change",{ target: {value} });

        //* Simulate submit
        wrapper.find("form").simulate("submit", { preventDefault(){} });

        //* Category has been called
        expect( setCategory ).toHaveBeenCalled();
        expect( setCategory ).toHaveBeenCalledTimes(1);
        expect( setCategory ).toHaveBeenCalledWith( expect.any(Function) );

        //* Clear input
        expect( wrapper.find("input").prop("value") ).toBe('');
    })
    

})
