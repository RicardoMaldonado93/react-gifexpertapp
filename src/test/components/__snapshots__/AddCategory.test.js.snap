// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for 'AddCategory' should render correctly component 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    style={
      Object {
        "opacity": "0.01",
        "position": "absolute",
      }
    }
  />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
